File: source/cs320_concept_programming_language_2014_fall/sec12/polyset.hs

module PolySet where

class IsSet s where
  empty :: s a
  insert :: Ord a => s a -> a -> s a
  contains :: (Ord a, Eq a) => a -> s a -> Bool
  
  inserts :: Ord a => s a -> [a] -> s a
  inserts s (x:xs) = inserts (insert s x) xs
  inserts s []     = s

data SetList a =
    Empty
  | Set [a]
  deriving Show   

instance IsSet SetList where
  empty = Empty

  insert (Set l) x = Set (x:l)
  insert (Empty) x = Set [x]

  contains x (Empty) = False
  contains x (Set l) = x `elem` l

data SetTree a =
    Leaf
  | Node a (SetTree a) (SetTree a)
  deriving Show   

instance IsSet SetTree where
  empty = Leaf

  insert (Leaf       ) x = Node x Leaf Leaf
  insert (Node y s s') x =
    if x < y then 
      Node y (insert s x) s' 
    else
      Node y s (insert s' x)

  contains x (Leaf       ) = False
  contains x (Node y s s') =
    if y == x then
      True
    else if x < y then
      contains x s
    else
      contains x s'

get :: (IsSet s) => s a
get = empty

both :: (Ord a, IsSet s1, IsSet s2) => s1 a -> s2 a -> a -> Bool
both s1 s2 e = (contains e s1) && (contains e s2)


main :: IO ()
main = do 
       let s1 = empty :: SetList int
           s2 = insert s1 1
           s3 = insert s2 2
           b1 = contains 2 s3
           b2 = contains 3 s3
       putStrLn (show b1)
       putStrLn (show b2)

       let
           s21 = empty :: SetTree int
           s22 = insert s21 1
           s23 = insert s22 3
           b21 = contains 3 s23
           b22 = contains 4 s23

       putStrLn (show b21)
       putStrLn (show b22)

       let
          b1 = both s3 s23 1
          b2 = both s3 s23 2
       putStrLn (show b1)
       putStrLn (show b2)

File: past_class/cs320_concept_programming_language_2014_fall/sec12/polyset.py



def both1(s1, s2, e):
    if (contains(s1, e) and contains(s2, e)):
        return True
    else:
        return False

def both2(s1, s2, e):
    if (s1.contains(e) and s2.contains(e)):
        return True
    else:
        return False

# ******************************************* #

def both(s1, s2, e):
    if (s1.get_type().contains(s1, e) and s2.get_type().contains(s2, e)):
        return True
    else:
        return False

class SetList(object):
    @staticmethod
    def empty():
        return SetList()

    @staticmethod
    def insert(s, e):
        if (s.m_lst == None):
            return SetList([e])
        else:
            return SetList([e] + s.m_lst)

    @staticmethod
    def contains(s, e):
        return e in s.m_lst

    def __init__(self, lst=None):
        self.m_lst = lst

    def get_type(self):
        return SetList

s1 = SetList.empty()
s2 = SetList.insert(s1, 1)
s3 = SetList.insert(s2, 2)

b = SetList.contains(s3, 2)
print b
b = SetList.contains(s3, 3)
print b

class SetTree(object):
    @staticmethod
    def empty():
        return SetTree()

    @staticmethod
    def insert(s, e):
        if (s.m_node == None):
            return SetTree((e, SetTree(), SetTree()))
        else:
            node = s.m_node
            if e < node[0]:
                sl = SetTree.insert(node[1], e)
                return SetTree((node[0], sl, node[2]))
            else:
                sr = SetTree.insert(node[2], e)
                return SetTree((node[0], node[1], sr))

    @staticmethod
    def contains(s, e):
        node = s.m_node
        if (node == None):
            return False
        elif e == node[0]:
            return True
        elif e < node[0]:
            return SetTree.contains(node[1], e)
        else:
            return SetTree.contains(node[2], e)

    def __init__(self, node=None):
        self.m_node = node

    def get_type(self):
        return SetTree

s21 = SetTree.empty()
s22 = SetTree.insert(s21, 1)
s23 = SetTree.insert(s22, 3)

b = SetTree.contains(s23, 3)
print b
b = SetTree.contains(s23, 4)
print b

b = both(s3, s23, 1)
print b
b = both(s3, s23, 2)
print b















